File: client/src/constants/fishing.js
var g_FISHING = {};
g_FISHING.FISH_RATES = {1:{FISH_NUM_MIN:5,FISH_NUM_MAX:7,FISH_RATE:{F2:45,F3:35,F4:20}},5:{FISH_NUM_MIN:7,FISH_NUM_MAX:9,FISH_RATE:{F2:40,F3:35,F4:25}},10:{FISH_NUM_MIN:9,FISH_NUM_MAX:12,FISH_RATE:{F2:30,F3:40,F4:30}},15:{FISH_NUM_MIN:12,FISH_NUM_MAX:15,FISH_RATE:{F2:25,F3:40,F4:35}}};
g_FISHING.FISHING_REWARD = {F3:{40:[{reward:{P47:1},rate:100},{reward:{P48:1},rate:50},{reward:{P49:1},rate:25}],70:[{reward:{P47:1},rate:100},{reward:{P48:1},rate:50},{reward:{P49:1},rate:25},{reward:{P50:1},rate:13}],100:[{reward:{P48:1},rate:100},{reward:{P49:1},rate:50},{reward:{P50:1},rate:25}]},F4:{35:[{reward:{P47:1},rate:100},{reward:{P48:1},rate:50},{reward:{P49:1},rate:25},{reward:{P50:1},rate:13}],65:[{reward:{P48:1},rate:100},{reward:{P49:1},rate:50},{reward:{P50:1},rate:25}],100:[{reward:{P48:1},rate:100},{reward:{P49:1},rate:50},{reward:{P50:1},rate:25},{reward:{P53:1},rate:13}]}};
g_FISHING.FISHING_MINIGAME_BAR = [{TYPE:'GREY',AREA_MIN:30,AREA_MAX:35,APPEAR_TIME:5,SLIDER_SPEED:80,GFX:'item_event03_weight_01'},{TYPE:'BRONZE',AREA_MIN:25,AREA_MAX:30,APPEAR_TIME:6,SLIDER_SPEED:100,GFX:'item_event03_weight_02'},{TYPE:'SILVER',AREA_MIN:15,AREA_MAX:20,APPEAR_TIME:7,SLIDER_SPEED:120,GFX:'item_event03_weight_03'},{TYPE:'GOLD',AREA_MIN:10,AREA_MAX:15,APPEAR_TIME:8,SLIDER_SPEED:140,GFX:'item_event03_weight_04'}];
g_FISHING.FISHING_MINIGAME_BAR_RATE = {0:[25,25,25,25],20:[20,30,25,25],50:[20,20,30,30],80:[15,25,30,30],120:[15,20,30,35]};
g_FISHING.FISH_WEIGHT = {GOLD:{F1:{MIN:1.6,MAX:2.0},F2:{MIN:5.5,MAX:6.3},F3:{MIN:7.5,MAX:8.3},F4:{MIN:9.5,MAX:10.3}},BRONZE:{F1:{MIN:0.6,MAX:0.9},F2:{MIN:3.5,MAX:4.3},F3:{MIN:5.5,MAX:6.3},F4:{MIN:7.5,MAX:8.3}},SILVER:{F1:{MIN:1.0,MAX:1.5},F2:{MIN:4.5,MAX:5.3},F3:{MIN:6.5,MAX:7.3},F4:{MIN:8.5,MAX:9.3}},GREY:{F1:{MIN:0.2,MAX:0.5},F2:{MIN:2.5,MAX:3.3},F3:{MIN:4.5,MAX:5.3},F4:{MIN:6.5,MAX:7.3}}};
g_FISHING.FISH_REWARD = {F1:{1:{GOLD:-1,EXP:25},21:{GOLD:-1,EXP:40},51:{GOLD:-1,EXP:60},81:{GOLD:-1,EXP:125},151:{GOLD:-1,EXP:250}},F2:{1:{GOLD:20,EXP:-1},21:{GOLD:40,EXP:-1},51:{GOLD:100,EXP:-1},81:{GOLD:200,EXP:-1},151:{GOLD:300,EXP:-1}},F3:{1:{GOLD:23,EXP:26},21:{GOLD:46,EXP:53},51:{GOLD:115,EXP:132},81:{GOLD:230,EXP:265},151:{GOLD:345,EXP:397}},F4:{1:{GOLD:26,EXP:30},21:{GOLD:53,EXP:61},51:{GOLD:132,EXP:152},81:{GOLD:265,EXP:305},151:{GOLD:397,EXP:457}}};
g_FISHING.FEATURE_DROP_LIST = {rules:{AIRSHIP_PACK:{rate:1.0,quantity:1.0,dailyLimit:-1.0,dropItemID:'HT1'},ORDER_GET_REWARD:{rate:1.0,quantity:2.0,dailyLimit:-1.0,dropItemID:'HT2'},ORDER_GET_REWARD_ORDER_DAILY_FREE:{rate:1.0,quantity:1.0,dailyLimit:-1.0,dropItemID:'HT1'},AIRSHIP_DELIVERY:{rate:1.0,quantity:5.0,dailyLimit:-1.0,dropItemID:'HT3'},TRUCK_PACK:{rate:1.0,quantity:1.0,dailyLimit:-1.0,dropItemID:'HT1'},TRUCK_DELIVERY:{rate:1.0,quantity:2.0,dailyLimit:-1.0,dropItemID:'HT3'},ORDER_GET_REWARD_ORDER_DAILY_PAID:{rate:1.0,quantity:3.0,dailyLimit:-1.0,dropItemID:'HT2'},MINE_START:{rate:1.0,quantity:3.0,dailyLimit:-1.0,dropItemID:'HT2'}},dropItemID:'HT1'};

File: tools/ExportExcel/src/exportexcel/sheet/Fishing.java
package exportexcel.sheet;

import data.*;
import exportexcel.Log;
import exportexcel.Util;
import org.apache.poi.ss.usermodel.Row;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.TreeMap;

public class Fishing extends ParseWorkbook {

    public FishingInfo fishingInfo = new FishingInfo();

    public Fishing(String inputName) throws Exception {
        super(inputName);
    }

    @Override
    public void handle() {
        Define.parseMiscInfo(parseSheetRow("Misc info"));
        parseFishRate(parseSheetRow("Fish Rate"));
        parseFishingReward(parseSheetRow("Reward Default"));
        parseMinigameBar(parseSheetRow("Minigame bar"));
        parseMinigameBarRate(parseSheetRow("Minigame bar Rate"));
        parseFishWeight(parseSheetRow("Fish Weight"));
        parseFishReward(parseSheetRow("Fish Reward"));
        parseFeatureDrop("Feature Drop");
        addConstInfo(fishingInfo, null);
    }

    private void parseFishRate(ParseSheetRow ps) {
        int maxRow = ps.sheet.getLastRowNum() + 1;

        for (int row = 1; row < maxRow; row++) {
            FishingInfo.FishRate fishRate = new FishingInfo.FishRate();

            int turn = ps.getInt(row, "TURN");
            fishRate.FISH_NUM_MIN = ps.getInt(row, "FISH_NUM_MIN");
            fishRate.FISH_NUM_MAX = ps.getInt(row, "FISH_NUM_MAX");
            fishRate.FISH_RATE = ps.getMapItemNum(row, "FISH_RATE");

            fishingInfo.FISH_RATES.put(turn, fishRate);
        }
    }

    private void parseFishingReward(ParseSheetRow ps) {
        int numRow = ps.sheet.getLastRowNum() + 1;
        String currentFish = "";
        int currentMilestone = 0;

        for (int row = 1; row < numRow; row++) {
            if (ps.isEmptyCell(row, "FISH")) {

                if (ps.isEmptyCell(row, "MILESTONE")) {
                    //Log.debug("cell: ", ps.getMapItemNum(row, "REWARD"));
                    HashMap<String, Integer> mapItem = ps.getMapItemNum(row, "REWARD");
                    int rate = ps.getInt(row, "RATE");

                    FishingInfo.RewardDefault rewardDefault = new FishingInfo.RewardDefault(mapItem, rate);
                    fishingInfo.FISHING_REWARD.get(currentFish).get(currentMilestone).add(rewardDefault);
                } else {
                    // Log.debug("cell: ", ps.getInt(row, "MILESTONE"));
                    currentMilestone = ps.getInt(row, "MILESTONE");
                    if (!fishingInfo.FISHING_REWARD.get(currentFish).containsKey(currentMilestone))
                        fishingInfo.FISHING_REWARD.get(currentFish).put(currentMilestone, new ArrayList<FishingInfo.RewardDefault>());
                }
            } else {
                // Log.debug("cell: ", ps.getString(row, "FISH"));
                currentFish = ps.getString(row, "FISH");
                if (!fishingInfo.FISHING_REWARD.containsKey(currentFish))
                    fishingInfo.FISHING_REWARD.put(currentFish, new TreeMap<Integer, List<FishingInfo.RewardDefault>>());
            }
        }
    }

    private void parseMinigameBar(ParseSheetRow ps) {
        int maxRow = ps.sheet.getLastRowNum() + 1;

        for (int row = 1; row < maxRow; row++) {
            FishingInfo.MinigameBar minigameBar = new FishingInfo.MinigameBar();
            minigameBar.TYPE = ps.getString(row, "TYPE");
            minigameBar.AREA_MIN = ps.getInt(row, "AREA_MIN");
            minigameBar.AREA_MAX = ps.getInt(row, "AREA_MAX");
            minigameBar.APPEAR_TIME = ps.getInt(row, "APPEAR_TIME");
            minigameBar.SLIDER_SPEED = ps.getInt(row, "SLIDER_SPEED");
	    minigameBar.GFX = ps.getString(row,"GFX");
            fishingInfo.FISHING_MINIGAME_BAR.add(minigameBar);
        }
    }
    private void parseMinigameBarRate(ParseSheetRow ps) {
        int maxRow = ps.sheet.getLastRowNum() + 1;

        for (int row = 1; row < maxRow; row++) {
            int level = ps.getInt(row, "LEVEL");
            int[] rate = ps.getArrayInt(row, "RATE");
            if (rate.length < fishingInfo.FISHING_MINIGAME_BAR.size())
                throw new RuntimeException("size Rate: "+ rate.length +"< num Type of Minigame bar : "+  fishingInfo.FISHING_MINIGAME_BAR.size());
           fishingInfo.FISHING_MINIGAME_BAR_RATE.put(level,rate);
        }
    }
    private String tryParseString(String input) {
        if (input == "null" || input == null || input.isEmpty())
            return input;

        String itemId = ParseWorkbook.mapIdName.inverse().get(input.toLowerCase());
        if (itemId != null)
            return itemId;

        String defineNum = Define.defineToString(input);
        if (defineNum != null)
            return defineNum;

        return null;
    }

    private void parseFeatureDrop(String idSheet) {
        ParseSheetRow parseSheet = parseSheetRow(idSheet);
        int maxRow = parseSheet.sheet.getLastRowNum();
        this.fishingInfo.FEATURE_DROP_LIST = new FeatureDropItemInfo(Util.toItemId(Define.miscJson.get("FISHING_DEFAULT_DROP_ITEM").getAsString()));
        for (int r = 1; r <= maxRow; r++) {
            Row row = parseSheet.sheet.getRow(r);
            if (row == null)
                break;

            String featureActionId = parseSheet.isEmptyCell(r, "ACTION") ? "" : parseSheet.getString(r, "ACTION");
            String option = parseSheet.isEmptyCell(r, "OPTION") ? "" : parseSheet.getString(r, "OPTION");
            float rate = parseSheet.isEmptyCell(r, "RATE") ? 1 : parseSheet.getFloat(r, "RATE");
            int quantity = parseSheet.isEmptyCell(r, "QUANTITY") ? -1 : parseSheet.getInt(r, "QUANTITY");
            int dailyLimit = parseSheet.isEmptyCell(r, "DAILY LIMIT") ? -1 : parseSheet.getInt(r, "DAILY LIMIT");
            String dropItemId = parseSheet.isEmptyCell(r, "TARGET_FISHING_TOKEN") ? "" : Util.toItemId(parseSheet.getString(r, "TARGET_FISHING_TOKEN"));

            if (featureActionId.isEmpty())
                continue;

            String temp = tryParseString(option);
            if (temp != null)
                option = temp;

            this.fishingInfo.FEATURE_DROP_LIST.add(featureActionId, option, rate, quantity, dailyLimit, dropItemId);
        }
    }

    private void parseFishWeight(ParseSheetRow ps) {
        int numRow = ps.sheet.getLastRowNum() + 1;
        String currentType = "";

        for (int row = 1; row < numRow; row++) {
            if (ps.isEmptyCell(row, "TYPE")) {

                String fish = Util.toItemId(ps.getString(row, "FISH_NAME"));
                FishingInfo.FishWeight fishWeight = new FishingInfo.FishWeight();
                fishWeight.MIN = ps.getFloat(row, "MIN");
                fishWeight.MAX = ps.getFloat(row, "MAX");
                if (!fishingInfo.FISH_WEIGHT.get(currentType).containsKey(fish))
                    fishingInfo.FISH_WEIGHT.get(currentType).put(fish, fishWeight);
            } else {
                currentType = ps.getString(row, "TYPE");
                if (!fishingInfo.FISH_WEIGHT.containsKey(currentType))
                    fishingInfo.FISH_WEIGHT.put(currentType, new TreeMap<String, FishingInfo.FishWeight>());
            }
        }
    }

    private void parseFishReward(ParseSheetRow ps) {
        int numRow = ps.sheet.getLastRowNum() + 1;
        String currentFish = "";

        for (int row = 1; row < numRow; row++) {
            if (ps.isEmptyCell(row, "FISH")) {

                int groupLV =ps.getInt(row, "GROUP_LV");
                FishingInfo.FishReward fishReward = new FishingInfo.FishReward();
                fishReward.GOLD = ps.getInt(row, "GOLD");
                fishReward.EXP = ps.getInt(row, "EXP");
                if (!fishingInfo.FISH_REWARD.get(currentFish).containsKey(groupLV))
                    fishingInfo.FISH_REWARD.get(currentFish).put(groupLV, fishReward);
            } else {
                currentFish =  Util.toItemId(ps.getString(row, "FISH"));
                if (!fishingInfo.FISH_REWARD.containsKey(currentFish))
                    fishingInfo.FISH_REWARD.put(currentFish, new TreeMap<>());
            }
        }
    }
}